File: java/src/main/java/com/meteroid/api/Invoices.java
// this file is @generated
package com.meteroid.api;

import com.meteroid.MeteroidHttpClient;
import com.meteroid.Utils;
import com.meteroid.exceptions.ApiException;
import com.meteroid.models.Invoice;
import com.meteroid.models.InvoiceCustomPropertiesRequest;
import com.meteroid.models.InvoiceListResponse;

import okhttp3.HttpUrl;

import java.io.IOException;

public class Invoices {
    private final MeteroidHttpClient client;

    public Invoices(MeteroidHttpClient client) {
        this.client = client;
    }

    /** List invoices with optional filtering by customer, subscription, or status. */
    public InvoiceListResponse listInvoices() throws IOException, ApiException {

        return this.listInvoices(new InvoicesListInvoicesOptions());
    }

    /** List invoices with optional filtering by customer, subscription, or status. */
    public InvoiceListResponse listInvoices(final InvoicesListInvoicesOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/invoices");
        if (options.customerId != null) {
            url.addQueryParameter("customer_id", options.customerId);
        }
        if (options.subscriptionId != null) {
            url.addQueryParameter(
                    "subscription_id", Utils.serializeQueryParam(options.subscriptionId));
        }
        if (options.statuses != null) {
            Utils.addExplodedQueryParameter(url, "statuses", options.statuses);
        }
        if (options.einvoicingStatus != null) {
            url.addQueryParameter(
                    "einvoicing_status", Utils.serializeQueryParam(options.einvoicingStatus));
        }
        if (options.orderBy != null) {
            url.addQueryParameter("order_by", options.orderBy);
        }
        if (options.page != null) {
            url.addQueryParameter("page", Utils.serializeQueryParam(options.page));
        }
        if (options.perPage != null) {
            url.addQueryParameter("per_page", Utils.serializeQueryParam(options.perPage));
        }
        return this.client.executeRequest(
                "GET", url.build(), null, null, InvoiceListResponse.class);
    }

    /** Retrieve a single invoice with its payment transactions. */
    public Invoice getInvoiceById(final String invoiceId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/invoices/%s", invoiceId));
        return this.client.executeRequest("GET", url.build(), null, null, Invoice.class);
    }

    /**
     * Merge custom property values onto an invoice (send a key with `null` to remove it). Values
     * are validated against the tenant's `INVOICE` property definitions. Allowed at any status —
     * custom properties are external workflow metadata and stay editable after the invoice is
     * finalized.
     */
    public Invoice patchInvoiceCustomProperties(
            final String invoiceId,
            final InvoiceCustomPropertiesRequest invoiceCustomPropertiesRequest)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(
                                String.format("/api/v1/invoices/%s/custom-properties", invoiceId));
        return this.client.executeRequest(
                "PATCH", url.build(), null, invoiceCustomPropertiesRequest, Invoice.class);
    }

    /** Download the PDF document for an invoice. */
    public byte[] downloadInvoicePdf(final String invoiceId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/invoices/%s/download", invoiceId));
        return this.client.executeBinaryRequest("GET", url.build(), null, null);
    }

    /**
     * Recompute a draft invoice against current usage, credits, coupons and tax, and return it.
     * Drafts are also refreshed periodically in the background; use this to force it, e.g. after
     * ingesting late events. Rejected while a payment for the invoice is in progress or when the
     * invoice was merged into a consolidated parent.
     */
    public Invoice refreshInvoice(final String invoiceId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/invoices/%s/refresh", invoiceId));
        return this.client.executeRequest("POST", url.build(), null, null, Invoice.class);
    }

    /**
     * Download the structured e-invoice (EN 16931 XML) issued with an invoice. For Factur-X the
     * same XML is also embedded in the PDF.
     */
    public byte[] downloadInvoiceXml(final String invoiceId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/invoices/%s/xml", invoiceId));
        return this.client.executeBinaryRequest("GET", url.build(), null, null);
    }
}

File: go/invoices.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// InvoicesListInvoicesOptions carries the query and header parameters of
// Invoices.ListInvoices.
//
// Optional parameters are pointers; leave them nil to omit them.
type InvoicesListInvoicesOptions struct {
	// Filter by customer ID or alias
	CustomerId *string

	SubscriptionId *SubscriptionId

	Statuses []InvoiceStatus

	// Sort order. Format: `column.direction`. Allowed columns: `invoice_number`, `customer_name`, `amount`, `invoice_date`, `status`, `payment_status`. Direction: `asc` or `desc`. Default: `invoice_date.desc`.
	OrderBy *string

	// Page number (0-indexed)
	Page *int32

	// Number of items per page
	PerPage *int32
}

// Invoices groups the invoices operations of the Meteroid API.
type Invoices struct {
	client *Client
}

// List invoices with optional filtering by customer, subscription, or status.
func (a *Invoices) ListInvoices(ctx context.Context, options *InvoicesListInvoicesOptions) (*InvoiceListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/invoices")

	if options != nil {
		if options.CustomerId != nil {
			req.SetQueryParam("customer_id", *options.CustomerId)
		}
		if options.SubscriptionId != nil {
			req.SetQueryParam("subscription_id", string(*options.SubscriptionId))
		}

		if len(options.Statuses) > 0 {
			for _, item := range options.Statuses {
				req.AddQueryParam("statuses", string(item))
			}
		}
		if options.OrderBy != nil {
			req.SetQueryParam("order_by", *options.OrderBy)
		}
		if options.Page != nil {
			req.SetQueryParam("page", formatInt(int64(*options.Page)))
		}
		if options.PerPage != nil {
			req.SetQueryParam("per_page", formatInt(int64(*options.PerPage)))
		}
	}

	var out InvoiceListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Retrieve a single invoice with its payment transactions.
func (a *Invoices) GetInvoiceById(ctx context.Context, invoiceId string) (*Invoice, error) {
	req := newRequest(http.MethodGet, "/api/v1/invoices/{invoice_id}")
	req.SetPathParam("invoice_id", invoiceId)

	var out Invoice
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Merge custom property values onto an invoice (send a key with `null` to remove it).
// Values are validated against the tenant's `INVOICE` property definitions. Allowed at any
// status — custom properties are external workflow metadata and stay editable after the invoice
// is finalized.
func (a *Invoices) PatchInvoiceCustomProperties(ctx context.Context, invoiceId string, invoiceCustomPropertiesRequest InvoiceCustomPropertiesRequest) (*Invoice, error) {
	req := newRequest(http.MethodPatch, "/api/v1/invoices/{invoice_id}/custom-properties")
	req.SetPathParam("invoice_id", invoiceId)

	req.SetJSONBody(invoiceCustomPropertiesRequest)

	var out Invoice
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Download the PDF document for an invoice.
func (a *Invoices) DownloadInvoicePdf(ctx context.Context, invoiceId string) ([]byte, error) {
	req := newRequest(http.MethodGet, "/api/v1/invoices/{invoice_id}/download")
	req.SetPathParam("invoice_id", invoiceId)

	return a.client.executeBinary(ctx, req)
}
